Language file support for cip-51 and Cx51

gtackett opened the pull request:
- added cip-51, and Cx51 to ldefs. 
- In 8051_main.sinc, created conditionalized address spaces, registers, etc., for this device based on CIP51 being defined

jobermayr commented:
```
diff --git a/Ghidra/Processors/8051/certification.manifest b/Ghidra/Processors/8051/certification.manifest
index ab3f5280a2..5dac87b24b 100644
--- a/Ghidra/Processors/8051/certification.manifest
+++ b/Ghidra/Processors/8051/certification.manifest
@@ -14,6 +14,8 @@ data/languages/8051.pspec||GHIDRA||||END|
 data/languages/8051.slaspec||GHIDRA||||END|
 data/languages/8051_archimedes.cspec||GHIDRA||||END|
 data/languages/8051_main.sinc||GHIDRA||||END|
+data/languages/cip-51.slaspec||GHIDRA||||END|
+data/languages/keil-cx51.cspec||GHIDRA||||END|
 data/languages/mx51.cspec||GHIDRA||||END|
 data/languages/mx51.pspec||GHIDRA||||END|
 data/languages/mx51.sinc||GHIDRA||||END|
```

gtackett commented:
> Do you know how I can resolve this?

Sorry, I'm running Ghidra on a Mac. I have a Windows VM but would rather not install Ghidra and its build tools there as well.

gtackett commented:
> Trying to build this branch results in:
> 
> ```
> FAILURE: Build failed with an exception.
> 
> * Where:
> Script 'C:\<...>\Ghidra_custom_CIP-51_support_2025-11\gradle\support\ip.gradle' line: 160
> 
> * What went wrong:
> Execution failed for task ':8051:ip'.
> > No IP found for C:\<...>\Ghidra_custom_CIP-51_support_2025-11\Ghidra\Processors\8051\data\languages\cip-51.slaspec in module: C:\<...>\Ghidra_custom_CIP-51_support_2025-11\Ghidra\Processors\8051. Expression: (ip != null). Values: ip = null
> ```
> 
On the command line for gradle[w] you can disable the check for the IP (intellectual property) clause. This works for me:
`gradle buildGhidra -x ip`

gtackett commented:
Oops. Didn't mean to close the pull request.

Pieter12345 commented:
> > > > Trying to build this branch results in:
> > > > ```
> > > > FAILURE: Build failed with an exception.
> > > > 
> > > > * Where:
> > > > Script 'C:\<...>\Ghidra_custom_CIP-51_support_2025-11\gradle\support\ip.gradle' line: 160
> > > > 
> > > > * What went wrong:
> > > > Execution failed for task ':8051:ip'.
> > > > > No IP found for C:\<...>\Ghidra_custom_CIP-51_support_2025-11\Ghidra\Processors\8051\data\languages\cip-51.slaspec in module: C:\<...>\Ghidra_custom_CIP-51_support_2025-11\Ghidra\Processors\8051. Expression: (ip != null). Values: ip = null
> > > > ```
> > > 
> > > 
> > > On the command line for gradle[w] you can disable the check for the IP (intellectual property) clause. This works for me: `gradle buildGhidra -x ip`
> > 
> > 
> > A git diff was posted in a [reply above](https://github.com/NationalSecurityAgency/ghidra/pull/8130#issuecomment-3507259015). Applying that got me to the build error that you commented on today. Note that I've been able to build this branch somewhere last year on the same system. I do not know which change caused my latest building issues, if it should build in the current state.
> 
> Have you actually tried building with the `-x ip` on your gradle command?
> 
> `-x ip` should be exactly what you need to get past the error text you got:
> 
> > No IP found for C:<...>\Ghidra_custom_CIP-51_support_2025-11\Ghidra\Processors\8051\data\languages\cip-51.slaspec in module: C:<...>\Ghidra_custom_CIP-51_support_2025-11\Ghidra\Processors\8051. Expression: (ip != null). Values: ip = null

I have. Either applying the git diff in the comment above or using the `-x ip` argument gets me past this error. In both cases, I run into the new error that I commented above. Do you know how to solve this? It seems like it cannot find one header file from my Visual Studio installation that is present in my file system. I have very little knowledge about the building process of this project.

gtackett commented:
> I have. Either applying the git diff in the comment above or using the `-x ip` argument gets me past this error. In both cases, I run into the new error that I commented above. Do you know how to solve this? It seems like it cannot find one header file from my Visual Studio installation that is present in my file system. I have very little knowledge about the building process of this project.

My mistake! I thought you were still getting the error about IP.

The other issue is one I can’t reproduce without using up a good bit of my time, and space on my Windows VM, to install Ghidra and its build dependencies and tools.  I’m willing to do that when time allows, if someone else can’t help.

ABratovic commented:
@Pieter12345 I believe your PDB build errors are not related to this pull request.
As per Ghidra/Features/PDB/src/pdb/README.txt you need to retarget the pdb.sln to match the VS2019 and SDK you have installed.

```
The PDB.EXE executable delivered with Ghidra has been built with Visual Studio 2017 and has a
dependency on the msdia140.dll in order to execute and process PDB files.  

If it is necessary to rebuild the PDB.exe using a newer version of the SDK, this can be done by 
using the included Visual Studio project to rebuild the executable.  After opening the project
solution file within Visual Studio simply re-target the solution to a new version of the SDK. 
This can be accomplished by opening the solution file within Visual Studio, right-clicking on 
the Solution 'pdb' node and selecting the "Re-target solution" menu item.  After re-targeting 
the solution simply rebuild it to produce a new executable (Ghidra/Features/PDB/os/win64/pdb.exe).
You may also need to adjust the solution configuration to build the pdb project when the 
solution is built. 
NOTES:
Registration of DIA SDK DLL required! (see docs/README_PDB.html file).

```

ABratovic commented:
@Pieter12345
BTW, if all you want to do is reverse the C8051F342 MCU, and not yet needing to modify any of the Java code, it may not be necessary to pull patches and build Ghidra from source. 

Simply unzip the prebuilt ghidra_11.4.2_PUBLIC and copy the modified 8051 Processor files from 
https://github.com/gtackett/ghidra/tree/language_files/Ghidra/Processors/8051/data/languages

If you further modify any of the files in this directory you can delete the *.sla files that Ghidra creates to be sure they get recreated on startup and you're ready to go pull your 8051 binary apart :)

gtackett committed:
Add 8051 CIP-51 (Silicon Labs) processor support and Keil C51 compiler specification

- Add CIP51 variant support to 8051_main.sinc with extended SFR addressing (2-byte)
- Add IFLASH and XRAM address spaces for CIP-51 memory model
- Add cip-51.slaspec for CIP-51 processor definition
- Add keil-cx51.cspec for Keil C51 compiler calling conventions

gtackett commented:
@jobermayr 
> ... Can you please also include IP fix ([#8130 (comment)](https://github.com/NationalSecurityAgency/ghidra/pull/8130#issuecomment-3507259015))

I think the idea is that the IP check is supposed to be disabled on the command line at build time by the person running the build, rather than disabled by config file? I could be wrong, though. Maybe @ryanmkurtz or one of the other Ghidra team members can chime in. Ryan initially told me how to disable this but I seem to remember that such a change to the official codebase wasn't desired?

GhidorahRex commented:
Under section D.6 of the [Github Terms of Service](https://docs.github.com/en/site-policy/github-terms/github-terms-of-service#d-user-generated-content), the content added is licensed under the same terms as the license in the repository unless you specify a license agreement separate to that. We have rejected pull-requests in the past for having an incompatible license.

Since this pull-request doesn't specify a license, it is assumed to be under the Ghidra (Apache2) license.

ABratovic commented:
@GhidorahRex 
To support the Silicon Labs C805121x with banked memory, instead of editing 8051_main.sinc and 
adding all the ```defined(CIP51)``` in this PR, I found all that is needed is a cip-51.slaspec file with
```
@define MCS51 ""
@define PTRSIZE 3

@include "8051_main.sinc"
```

I tidied up cip-51.pspec (found it in pull 7320) and added PSBANK 0xB1, and created a separate for now ldefs file.

Extract the three files in [cip-51.zip](https://github.com/user-attachments/files/26995760/cip-51.zip) into the 8051 languages directory, load the Intel HEX file in [C8051F21x_BANKING.zip](https://github.com/user-attachments/files/26995860/C8051F21x_BANKING.zip) and select 8051 variant cip-51. 
Goto 18000 and press D.
Note the LCALL is incorrect.
<img width="647" height="523" alt="bank1" src="https://github.com/user-attachments/assets/a4da46c4-2c0b-4478-9e8c-e2d743702288" />

In 8051_main.sinc, replace 
```
@else
:LCALL Addr16 is $(GROUP1) & ophi=1 & oplo=2; Addr16	{  ret:$(PTRSIZE) = inst_next; push16(ret); call Addr16; }
:LJMP Addr16  is $(GROUP1) & ophi=0 & oplo=2; Addr16	{ goto Addr16; }
@endif
```
with
```
@else
@if PTRSIZE == "3"
CallDest: addr16 is addr16 {
    local addr:3 = addr16; 
    if (addr < 0x8000) goto <dada>;
    addr = addr ^ (inst_start & 0xFF0000);
<dada>
    export addr;
}
:LCALL CallDest is $(GROUP1) & ophi=1 & oplo=2; CallDest { ret:$(PTRSIZE) = inst_next; push16(ret); call [CallDest]; }
:LJMP CallDest  is $(GROUP1) & ophi=0 & oplo=2; CallDest { goto [CallDest]; }
@else
:LCALL Addr16 is $(GROUP1) & ophi=1 & oplo=2; Addr16	{  ret:$(PTRSIZE) = inst_next; push16(ret); call Addr16; }
:LJMP Addr16  is $(GROUP1) & ophi=0 & oplo=2; Addr16	{ goto Addr16; }
@endif
@endif
```
restart Ghidra, reload C8051F21x_BANKING.HEX and goto 18000 and press D, then press D at 38000, 0 and 36.
Success.
<img width="608" height="638" alt="cip-51-success" src="https://github.com/user-attachments/assets/31f9fa35-4db9-4c59-a1a2-b5a7f3cf4753" />

gtackett commented:
### Hurrah!
Unless I find myself with an unexpected bounty of free time and perhaps with a little incentive to pick up a Silicon Labs development board, this likely will be my last contribution to the language files for the C8051Fxxx. I began this effort just as COVID was about to hit. It (and I) owe a lot to the encouragement and assistance of @ryanmkurtz.

Thanks everyone. It's been fun!

ABratovic commented:
@ryanmkurtz FYI, the files merged (and not merged) in https://github.com/NationalSecurityAgency/ghidra/commit/491bfe15c63d8340ff2d5fdff3477190b3899a02 are NOT what was submitted and peer reviewed prior to the force-pushes in January. Based on what WAS committed and what I've noted above, I would suggest reversing the commit and start again.

GhidorahRex commented:
@ABratovic The forced-push changes were just rebasing the original, as far as I can tell. What got merged looks like what is included in this ticket. Is there something that @gtackett missed here?

ABratovic commented:
> @ABratovic The forced-push changes were just rebasing the original, as far as I can tell. What got merged looks like what is included in this ticket. Is there something that @gtackett missed here?

The changes to ldef disappeared, cip-51.pspec is missing.
The sinc file is not the same as was submitted and the cspec contains a segment pcodeop that is not referenced anywhere in the sinc file. 

And as noted above, cip-51 support can be achieved simply by adding @define PTRSIZE 3, and creating a new SFRSIZE =2 when someone wants to add support for SFRPAGE

GhidorahRex commented:
@ABratovic Thanks for the clarification, I'll get a fix in. I can address the missing ldef entry and pspec file, clean up the cspec, and look into the fixes you have for the slaspec.

ABratovic commented:
There's pspec in my zip file above, and here's a tidied up cspec, I changed the whitespace so you can more easily file compare it to 8051.cspec. [keil-cx51.zip](https://github.com/user-attachments/files/27480674/keil-cx51.zip)

GhidorahRex commented:
Thanks @ABratovic

ABratovic commented:
And here's my sinc file with fix for banked lcall below 0x8000 and a few other changes you can cherry pick from.

[8051_main.zip](https://github.com/user-attachments/files/27481327/8051_main.zip)

ABratovic commented:
[jobermayer-8130-Language-file-support-for-cip-51-and-Cx51_Sept2025.patch](https://github.com/user-attachments/files/27506041/jobermayer-8130-Language-file-support-for-cip-51-and-Cx51_Sept2025.patch)

[8051_main_gtackett_Jan2025.zip](https://github.com/user-attachments/files/27506055/8051_main_gtackett_Jan2025.zip)

This sinc file and a patch file circa Sept 2025 including all the other relevant files preceeding the force-pushes I believe could be pretty close to what was originally submitted and peer reviewed in this pull request. Note the absence of all the cip-51 design comments at start of sinc file. I think the only comment needed, if any, is that the cip-51 has multiple SFR pages.
I believe the space RAM, IFLASH and possibly XRAM could be superfluous when CODE is defined with PTRSIZE=3, 
which also saves the user having to go into options when loading to select IFLASH for banked memory files like my 
Intel Hex example uploaded above https://github.com/NationalSecurityAgency/ghidra/pull/8130#issuecomment-4301599241.

Essentially every ```|| defined(CIP51)``` (other than SFR size=2 for which there is no supporting code as yet so is it really needed?) has been added to a line containing ```if defined(MCS51)``` making the CIP51 define redundant in my eyes.

I also noticed in this 2025 sinc file the missing define @GhidorahRex added after ```@if defined(MCS251) || defined(MX51)```  in https://github.com/NationalSecurityAgency/ghidra/commit/9e18f9439e54f13d31c9fb2b7de373415849150b was (possibly incorrectly for banked memory?) placed by gtackett after the ```@elif defined(MCS51)```. 
```
@if defined(MCS251) || defined(MX51)
Addr11: relAddr is aopaddr & adata [ relAddr = (inst_next $and 0xfff800)+(aopaddr*256)+adata; ]  { export *:1 relAddr; }
Addr16: addr is addr16 [ addr = (inst_next $and 0xff0000) + addr16; ] { export *:1 addr; }
@elif defined(MCS51) || defined(CIP51)
Addr11: relAddr is aopaddr & adata [ relAddr =(inst_next $and 0xf800)+(aopaddr*256)+adata; ]  { export *:1 relAddr; }
Addr16: addr16 is addr16 { export *:1 addr16; }
@elif defined(MCS80390)
Addr19: relAddr is aopaddr & adata [ relAddr =(inst_next $and 0xf80000)+(aopaddr*256*256)+adata; ]  { export *:1 relAddr; }
Addr24: addr24 is addr24 { export *:1 addr24; }
@endif
```
It was this missing define I noted above on Mar 31st that made me go looking for a better way to support a cip-51 with banked memory, and finding the sinc file already supported it with a simple ```@define PTRSIZE 3``` that is also applicable to many of the other 8051 derivatives with banked memory without needing a ```@define CIP51```.

ABratovic commented:
For anyone landing here wondering what happened to the mooted cip-51 support in v12.1 you'll need to wait until patch v12.1.1, or v12.2 is released.
The missing files and edits I noted and provided were added in https://github.com/NationalSecurityAgency/ghidra/commit/8a09a052583f60432b68bcadcdc5adfb5fbf16e5. The commit is dated May 8, but with the vagaries of repository mirroring and the likely holding back of internal commits until the internal nightly build passes, I don't recall seeing it appear in master until after v12.1 was released so was unable to advise it wasn't going to be in v12.1 :face_with_head_bandage:
